luajit: fix building for iOS and Android

kambala-decapitator opened the pull request:
### Summary
Changes to recipe:  **luajit/all**

#### Motivation

- Fixes building for iOS and Android
- forbids building for 32-bit platform on macOS: https://github.com/LuaJIT/LuaJIT/issues/664
- adds passing `CPPFLAGS` as `TARGET_CFLAGS` since upstream doesn't read this standard variable
- fixes producing .so symlinks when `ldconfig` is unavailable (e.g. building for Android on macOS) via upstream patch, but I had to adjust it a bit that it applies on 2.1.0-beta3
- fixes extension of shared lib when building for iOS via upstream patch, but I had to adjust it a bit that it applies on 2.1.0-beta3
- simplifies using Clang compiler
- adds option to enable GC64 mode. Keeping it off by default to match behavior of the source, but in later LuaJIT versions it's on by default.

#### Details

TODO: drop the last commits after #30447 is merged, there will also be conflicts

Followed the documentation at https://luajit.org/install.html#cross, tested building 2.1.0-beta3 for iOS on macOS and for Android on macOS and Linux.

I will explain why env modification is done. Without it target flags (i.e. `host` in Conan terms) "taint" the host ones (`build` in Conan terms) which results in non-working executable that's meant to run on the `build` platform, it can be seen in the build log:

```
HOSTCC    host/minilua.o
clang  -O2 -fomit-frame-pointer -Wall   -O3 -isysroot /Applications/Xcode-15.4.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.5.sdk -arch arm64 -mios-version-min=12.0 -I. -DLUAJIT_OS=LUAJIT_OS_OSX -DLUAJIT_TARGET=LUAJIT_ARCH_arm64 -DLJ_ARCH_HASFPU=1 -DLJ_ABI_SOFTFP=0 -DLUAJIT_NO_UNWIND  -c -o host/minilua.o host/minilua.c

HOSTLINK  host/minilua
clang   -isysroot /Applications/Xcode-15.4.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.5.sdk -arch arm64 -mios-version-min=12.0   -o host/minilua host/minilua.o -lm 
```

Checking how the host executable is produced:

```
$(HOST_O): %.o: %.c
	$(E) "HOSTCC    $@"
	$(Q)$(HOST_CC) $(HOST_ACFLAGS) -c -o $@ $<


ASOPTIONS= $(CCOPT) $(CCWARN) $(XCFLAGS) $(CFLAGS)
CCOPTIONS= $(CCDEBUG) $(ASOPTIONS)
HOST_ACFLAGS= $(CCOPTIONS) $(HOST_XCFLAGS) $(TARGET_ARCH) $(HOST_CFLAGS)
```

first I came up with the following solution that overwrites those target flags with proper ones:

```python
xcrun_host = XCRun(self, sdk='macosx')
arch_host = {
    'x86_64': 'x86_64',
    'armv8': 'arm64',
}.get(str(self.settings_build.get_safe("arch")))
target_flag_host = f"{arch_host}-apple-macos{self.settings_build.get_safe("os.version", default='')}"
flags_host = f"'-isysroot \"{xcrun_host.sdk_path}\" -target {target_flag_host}'"

# and pass additional variables as make args:
[
    f"HOST_CFLAGS={flags_host}",
    f"HOST_LDFLAGS={flags_host}",
]
```

which results in proper executable:

```
HOSTCC    host/minilua.o
clang  -O2 -fomit-frame-pointer -Wall   -O3 -isysroot /Applications/Xcode-15.4.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.5.sdk -arch arm64 -mios-version-min=12.0 -I. -DLUAJIT_OS=LUAJIT_OS_OSX -DLUAJIT_TARGET=LUAJIT_ARCH_arm64 -DLJ_ARCH_HASFPU=1 -DLJ_ABI_SOFTFP=0 -DLUAJIT_NO_UNWIND -isysroot "/Applications/Xcode-15.4.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk" -target arm64-apple-macos -c -o host/minilua.o host/minilua.c
clang: warning: overriding '-mios-version-min=12.0' option with '-target arm64-apple-macos' [-Woverriding-t-option]


HOSTLINK  host/minilua
clang   -isysroot /Applications/Xcode-15.4.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.5.sdk -arch arm64 -mios-version-min=12.0  -isysroot "/Applications/Xcode-15.4.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk" -target arm64-apple-macos -o host/minilua host/minilua.o -lm   
clang: warning: overriding '-mios-version-min=12.0' option with '-target arm64-apple-macos' [-Woverriding-t-option]
```

But later I figured out that it's much easier to simply drop CFLAGS / LDFLAGS from the environment, as LuaJIT uses its own variables for host/target flags. And Android has quite a similar issue.

Also when building a _build_ binary in Android build on macOS, a workaround is required - passing `-isysroot` in HOST_CFLAGS and HOST_LDFLAGS. This looks like a common issue related to the NDK recipe, I've opened #26585 about that.

---
- [x] Read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md)
- [x] Checked that this PR is not a duplicate: [list of PRs by recipe](https://github.com/conan-io/conan-center-index/discussions/24240)
- [x] Tested locally with at least one configuration using a recent version of Conan

kambala-decapitator committed:
define clang compiler via variable instead of patching Makefile

kambala-decapitator committed:
fix building for iOS

kambala-decapitator committed:
fix building for Android 64-bit

valgur commented on recipes/luajit/all/conanfile.py:
```diff
@@ -83,18 +88,26 @@ def _patch_sources(self):
         args = [f"PREFIX={unix_path(self, self.package_folder)}"]
-        if is_apple_os(self) and self._macosx_deployment_target:
-            args.append(f"MACOSX_DEPLOYMENT_TARGET={self._macosx_deployment_target}")
+        if "clang" in str(self.settings.compiler):
+            args.append("DEFAULT_CC=clang")
```
This is incorrect and basically reproduces the existing sloppy handling of CC in the recipe. It hackily replaces the default `CC=gcc` with `clang`, but it ignores the exact compiler executable set in the Conan profile and won't work for any other compilers in general.

It should really be something like
```python
cc = AutotoolsToolchain(self).vars()["CC"]
args.append(f"CC={cc}")
```
Similarly, `HOST_CC` needs to be set to the `CC_FOR_BUILD` value from the toolchain vars. Something like https://github.com/conan-io/conan/issues/17602 would be very welcome here.

kambala-decapitator replied:
that's what iOS build docs suggest to do. So should I add this variable only for iOS then and revert the general compiler adjustment?

valgur commented:
I have not opened a PR yet, but most of the make args except `TARGET_FLAGS` set here seem to be necessary for cross-compilation in general: https://github.com/valgur/conan-center-index/blob/dev/recipes/luajit/all/conanfile.py#L53-L88

kambala-decapitator commented:
@valgur would it be ok to add another commit that fixes Android build (it builds on top of this change w.r.t. clearing CFLAGS / CXXFLAGS / LDFLAGS) or wait for this PR to be merged and then open another one?

valgur commented:
Would be ok by me, but I have no say in this as a regular contibutor.

kambala-decapitator committed:
raise on attempt to build for 32-bit platform on macOS

kambala-decapitator committed:
fix building for Android 32-bit

kambala-decapitator committed:
add upstream patch that fixes .so symlinks on systems without ldconfig

e.g. building for Android on macOS

the patch was adjusted to apply on 2.1.0-beta3

kambala-decapitator committed:
inject CPPFLAGS as TARGET_CFLAGS

kambala-decapitator committed:
add upstream patch that sets dylib extension also for iOS shared build

the patch was adjusted to apply on 2.1.0-beta3

kambala-decapitator commented on recipes/luajit/all/conanfile.py:
```diff
@@ -41,6 +42,15 @@ def configure(self):
+            raise ConanInvalidConfiguration(f"{self.ref} cannot be cross-built to a 32-bit platform on macOS, see https://github.com/LuaJIT/LuaJIT/issues/664")
+
     def validate(self):
         if self.settings.os == "Macos" and self.settings.arch == "armv8" and cross_building(self):
             raise ConanInvalidConfiguration(f"{self.ref} can not be cross-built to Mac M1. Please, try any version >=2.1")
```
to my understanding this should also go to `validate_build`, please let me know if I should adjust it in this PR

kambala-decapitator committed:
add option to enable GC64 mode

kambala-decapitator committed:
increase cmake_minimum_required in the test package

kambala-decapitator commented:
closing in favor of https://github.com/kambala-decapitator/conan-luajit
